EZP-28633: Implemented support for custom wildcard URLs

kmadejski opened the pull request:
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-28633](https://jira.ez.no/browse/EZP-28633) ([EZP-29532](https://jira.ez.no/browse/EZP-29532) / [EZP-29533](https://jira.ez.no/browse/EZP-29533))
| **Bug/Improvement**| no
| **New feature**    | yes, eZ Publish feature parity
| **Target version** | `7.5`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | yes

This PR introduces the missing implementation of `UrlWildcardRouter`. As agreed with @SylvainGuittard it contains only PHP API side, which should also allow reusing UrlWildcards configured in legacy. The UI part is planned for v3.x and is not included in this PR. 

`eZ/Publish/API/Repository/URLWildcardService::exists($url)` method has been introduced as a optimisation and a simple way to avoid using `eZ/Publish/API/Repository/URLWildcardService::translate($url)` during the new UrlWildcard creation.

### Configuration
To enable URLWildcards Router the following configuration is required:
```
ezpublish:
    url_wildcards:
        enabled: true
```


**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement unit tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.
- [x] Implement integration tests.

adamwojs commented on eZ/Publish/API/Repository/URLWildcardService.php:
```diff
@@ -75,4 +75,15 @@ public function loadAll($offset = 0, $limit = -1);
+     *
+     * @return \eZ\Publish\API\Repository\Values\Content\URLWildcard
+     */
+    public function lookup($url);
```
```suggestion
    public function lookup(string $url): URLWildcard;
```

adamwojs commented on eZ/Publish/Core/MVC/Symfony/Routing/UrlWildcardRouter.php:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    /**
+     * @param Request $request
```
Nitpick: Please use FQCN

adamwojs commented on eZ/Publish/Core/MVC/Symfony/Routing/UrlWildcardRouter.php:
```diff
@@ -0,0 +1,210 @@
+        $this->generator->setRequestContext($context);
+    }
+
+    public function getContext()
```
```suggestion
    public function getContext(): RequestContext
```

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlWildcard/Gateway.php:
```diff
@@ -49,4 +49,13 @@ abstract class Gateway
+     *
+     * @return array
+     */
+    abstract public function loadUrlWildcardByUrl($url);
```
```suggestion
    abstract public function loadUrlWildcardByUrl(string $url): array;
```

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlWildcard/Gateway/ExceptionConversion.php:
```diff
@@ -103,4 +103,22 @@ public function loadUrlWildcardsData($offset = 0, $limit = -1)
+     *
+     * @return array
+     */
+    public function loadUrlWildcardByUrl($url)
```
This method could be simplified to 

```php
        try {
            return $this->innerGateway->loadUrlWildcardByUrl($url);
        } catch (DBALException | PDOException $e) {
            throw new RuntimeException('Database error', 0, $e);
        }
```

adamwojs commented on eZ/Publish/Core/SignalSlot/URLWildcardService.php:
```diff
@@ -151,4 +151,19 @@ public function translate($url)
+     */
+    public function lookup($url)
+    {
+        // @todo implement LookupSignal
```
We are not emitting signals for "load" methods 😉

andrerom commented on eZ/Publish/Core/Persistence/Cache/UrlWildcardHandler.php:
```diff
@@ -0,0 +1,161 @@
+            [
+                'id' => $id,
+            ]
+        );
```
more of a personal preference, but when this fits on one line. But that said maybe more for @alongosz and others to have a say on what they prefer here, or check with symfony CS which we try to align with.

same below

kmadejski replied:
I've cleaned it up as you requested.

andrerom commented on eZ/Publish/Core/Persistence/Cache/UrlWildcardHandler.php:
```diff
@@ -0,0 +1,161 @@
+        $cacheItem->tag($this->getCacheTags([$urlWildcard]));
+        $this->cache->save($cacheItem);
+
+        $cacheItemBySource = $this->cache->getItem('ez-urlWildcard-source-' . $this->escapeForCacheKey($urlWildcard->sourceUrl));
```
afaik, no need to pre warm the cache for another key here
1. complicates tests
2. fills up cache with cache that might not be used

And last but not least, isth the logic here flawed? `sourceUrl` is typically a pattern right? while url sent to `lookup` is an actual url. So this prewarmed item will never be used .

andrerom commented on eZ/Publish/Core/MVC/Symfony/Routing/UrlWildcardRouter.php:
```diff
@@ -0,0 +1,210 @@
+     *
+     * @throws NotFoundException
+     */
+    protected function getUrlWildcard($requestedPath): APIUrlWildcard
```
these methods does not seem to add any value besides making the code harder to read, or?

kmadejski replied:
Indeed, removed them.

adamwojs commented on eZ/Publish/Core/MVC/Symfony/Routing/UrlWildcardRouter.php:
```diff
@@ -0,0 +1,210 @@
+            }
+        }
+
+        if (isset($this->logger)) {
```
This expression is always true. Ref: https://github.com/ezsystems/ezpublish-kernel/pull/2609/files#diff-f739df759a10b5a8df9a6c0930145fcbR65

adamwojs commented on eZ/Publish/Core/MVC/Symfony/Routing/UrlWildcardRouter.php:
```diff
@@ -0,0 +1,210 @@
+     *
+     * @throws NotFoundException
+     */
+    protected function getUrlWildcardTranslationResult($requestedPath): URLWildcardTranslationResult
```
Nitpick: Missing type hints

andrerom commented on eZ/Publish/Core/Persistence/Cache/UrlWildcardHandler.php:
```diff
@@ -0,0 +1,161 @@
+            ]
+        );
+
+        $cacheItem = $this->cache->getItem('ez-urlWildcard-all-' . $offset . '-' . $limit);
```
I'll rather drop caching load all, it can be a very large result. _So assuming we fix the design here there won't be a need to pre-emptive try to optimize a load all_

andrerom replied:
ref legacy comment above, maybe the answer is not to load all, but rather expose a ending to load all in a compact format that can be effectively used by regex matching.

Unless someone sees how we can do the match in sql _(picking first match by order)_, and cache the results.

andrerom commented on eZ/Publish/API/Repository/URLWildcardService.php:
```diff
@@ -75,4 +75,15 @@ public function loadAll($offset = 0, $limit = -1);
+     *
+     * @return \eZ\Publish\API\Repository\Values\Content\URLWildcard
+     */
+    public function lookup($url);
```
I'm unsure why we need a new API.
The load all issue with `translate` is a implementation issue, and not a design issue right?

As in, we can:
- Add new method in SPI to do the lookup more effectively and cache that
- Use that in translate()

That said, as discussed on slack, even legacy loads all kind of, as in it generates a set of php files based that it caches to do the matching in pure php, but quite effectively:
https://github.com/ezsystems/ezpublish-legacy/blob/master/kernel/classes/ezurlwildcard.php#L217

kmadejski replied:
That's right. I've removed that new API method, kept it only for SPI handler. The method is used only in `UrlWildcardService::translate()`.

> Add new method in SPI to do the lookup more effectively and cache that

The SPIs `lookup` method is cached, please refer to: https://github.com/ezsystems/ezpublish-kernel/pull/2609/files#diff-485676a9318fdeefc018635a35f8a7faR77

I'll look for some good approach for `loadAll()`.

andrerom commented on eZ/Publish/Core/Persistence/Cache/UrlWildcardHandler.php:
```diff
@@ -0,0 +1,161 @@
+        $cacheItem->tag($this->getCacheTags([$urlWildcard]));
+        $this->cache->save($cacheItem);
+
+        $cacheItemById = $this->cache->getItem('ez-urlWildcard-id-' . $urlWildcard->id);
```
same, no need to pre warm things

kmadejski replied:
Alright, I understand, I've removed all pre-warmings here and there.

SylvainGuittard commented:
Thanks @kmadejski for this contribution.
Your PR is against master. Will this be available on 2.5.x?

andrerom commented:
@SylvainGuittard It's meant for 2.5, there is just no brach for it yet.

andrerom commented on eZ/Bundle/EzPublishCoreBundle/DependencyInjection/Compiler/UrlWildcardRouterPass.php:
```diff
@@ -0,0 +1,48 @@
+
+        if ($container->getParameter('ezpublish.url_wildcards.enabled') === true) {
+            $definition = new Definition();
+            $definition->setClass($container->getParameter('ezpublish.urlwildcard_router.class'));
```
I guess we can drop this given we move towards just using class name as service name, or wdyt @alongosz ?

alongosz replied:
Definitely we should not use `.class`. It's been deprecated in Symfony 2.7 or 2.8 and dropped in 3.x (AFAIR).

andrerom commented on eZ/Publish/Core/MVC/Symfony/Routing/UrlWildcardRouter.php:
```diff
@@ -0,0 +1,169 @@
+        // redirect if URLWildcard is configured as forward
+        if ($urlWildcardTranslationResult->forward) {
+            $params = [
+                '_controller' => 'Symfony\Bundle\FrameworkBundle\Controller\RedirectController::urlRedirectAction',
```
possible to use `::class` here?

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlWildcard/Gateway/DoctrineDatabase.php:
```diff
@@ -162,4 +162,33 @@ public function loadUrlWildcardsData($offset = 0, $limit = -1)
+     *
+     * @return array
+     */
+    public function loadUrlWildcardBySourceUrl(string $sourceUrl)
```
```suggestion
    public function loadUrlWildcardBySourceUrl(string $sourceUrl): array
```

adamwojs replied:
```suggestion
    public function loadUrlWildcardBySourceUrl(string $sourceUrl): array
```

andrerom commented on eZ/Publish/Core/Persistence/Cache/UrlWildcardHandler.php:
```diff
@@ -0,0 +1,148 @@
+
+        $this->persistenceHandler->urlWildcardHandler()->remove($urlWildcard->id);
+
+        // need to clear lists of UrlWildcards cached due to loadAll()
```
I guess this is not needed anymore now that loadAll does not cache anymore?

andrerom commented on eZ/Publish/Core/Persistence/Cache/UrlWildcardHandler.php:
```diff
@@ -0,0 +1,148 @@
+        $cacheItem = $this->cache->getItem('ez-urlWildcard-id-' . $id);
+
+        if ($cacheItem->isHit()) {
+            if (($return = $cacheItem->get()) === self::NOT_FOUND) {
```
wrong copy paste from translate right? as in on id lookup we don't need to cache missing items I would assume?

kmadejski replied:
Indeed, good catch, thanks!

andrerom commented on eZ/Publish/Core/Persistence/Cache/UrlWildcardHandler.php:
```diff
@@ -0,0 +1,148 @@
+                ->expiresAfter(30)
+                ->tag(['urlWildcard-notFound']);
+            $this->cache->save($cacheItem);
+            throw $e;
```
I know we already do this, but I'm wondering if it might be better to wrap it in order to make sure line numbers and trace is kept correct on the original exception _(assuming the new throw tampers with that)_.

kmadejski replied:
Yes, that's a good approach. I improved it.

adamwojs commented on eZ/Publish/Core/MVC/Symfony/Routing/UrlWildcardRouter.php:
```diff
@@ -0,0 +1,169 @@
+        $this->wildcardService = $wildcardService;
+        $this->generator = $generator;
+        $this->requestContext = $requestContext;
+        $this->logger = $logger;
```
Can we inject $logger via setter?

alongosz replied:
> Can we inject $logger via setter?

Why? I consider setters a bad practice.

kmadejski replied:
@adamwojs yes, introduced it in https://github.com/ezsystems/ezpublish-kernel/pull/2609/commits/68791bac3821e41dac5af1ea26f23f2c4185f63c.

adamwojs commented on eZ/Publish/Core/Persistence/Cache/UrlWildcardHandler.php:
```diff
@@ -0,0 +1,148 @@
+    /**
+     * Constant used for storing not found results for lookup().
+     */
+    const NOT_FOUND = 0;
```
```suggestion
    private const NOT_FOUND = 0;
```

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlWildcard/Gateway.php:
```diff
@@ -49,4 +49,13 @@ abstract class Gateway
+     *
+     * @return array
+     */
+    abstract public function loadUrlWildcardBySourceUrl(string $sourceUrl);
```
```suggestion
    abstract public function loadUrlWildcardBySourceUrl(string $sourceUrl): array;
```

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlWildcard/Gateway/ExceptionConversion.php:
```diff
@@ -103,4 +103,20 @@ public function loadUrlWildcardsData($offset = 0, $limit = -1)
+     *
+     * @return array
+     */
+    public function loadUrlWildcardBySourceUrl(string $sourceUrl)
```
```suggestion
    public function loadUrlWildcardBySourceUrl(string $sourceUrl): array
```

andrerom commented on eZ/Publish/Core/Repository/URLWildcardService.php:
```diff
@@ -82,14 +78,11 @@ public function create($sourceUrl, $destinationUrl, $forward = false)
-                    'Pattern already exists'
-                );
-            }
+        if ($this->urlWildcardHandler->exists($this->cleanPath($sourceUrl))) {
```
Exists is prone to race conditions if people for some reason end up doing exists + load call.
Would it make sense rather do a translate() here and enhance the exception with the info so it becomes `Pattern already exists and currently points to <target>`

In order to be able to drop `exists()` 😅

andrerom replied:
EDIT: I see exists has different semantic now, it only looks for exact match and not wildcards.

Maybe rename to `exactSourceUrlExists()`, aka something more specific? _(or something better, check in the office what others think)_

either way also phpdoc should better reflect this.

kmadejski replied:
I changed the method name to the proposed one. I hope the intentions are more clear now.

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlWildcard/Handler.php:
```diff
@@ -113,4 +114,107 @@ public function loadAll($offset = 0, $limit = -1)
+     *
+     * @return string
+     */
+    private function compile($sourceUrl): string
```
```suggestion
    private function compile(string $sourceUrl): string
```

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlWildcard/Handler.php:
```diff
@@ -113,4 +114,107 @@ public function loadAll($offset = 0, $limit = -1)
+     *
+     * @return string
+     */
+    private function substitute($destinationUrl, array $values): string
```
```suggestion
    private function substitute(string $destinationUrl, array $values): string
```

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlWildcard/Handler.php:
```diff
@@ -113,4 +114,107 @@ public function loadAll($offset = 0, $limit = -1)
+     */
+    private function substitute($destinationUrl, array $values): string
+    {
+        preg_match_all('(\{(\d+)\})', $destinationUrl, $matches);
```
Please introduce const representing this regular expression

adamwojs commented on eZ/Publish/Core/Repository/URLWildcardService.php:
```diff
@@ -125,19 +118,31 @@ public function create($sourceUrl, $destinationUrl, $forward = false)
+     *
+     * @return string
+     */
+    protected function cleanPath($path): string
```
```suggestion
    private function cleanPath(string $path): string
```

adamwojs commented on eZ/Publish/Core/MVC/Symfony/Routing/UrlWildcardRouter.php:
```diff
@@ -0,0 +1,169 @@
+    /**
+     * Not supported. Please use matchRequest() instead.
+     *
+     * @param $pathinfo
```
```suggestion
     * @param string $pathinfo
```

adamwojs commented on eZ/Publish/Core/Persistence/Cache/UrlWildcardHandler.php:
```diff
@@ -0,0 +1,148 @@
+     * @param string $sourceUrl
+     * @return bool
+     */
+    public function exists(string $sourceUrl): bool
```
Nitpick: Could you please move this method before `getCacheTags`?

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlWildcard/Handler.php:
```diff
@@ -113,4 +114,107 @@ public function loadAll($offset = 0, $limit = -1)
+     *
+     * @return string|null
+     */
+    private function match($url, $wildcard): ?string
```
```suggestion
    private function match(string $url, array $wildcard): ?string
```

adamwojs commented on eZ/Publish/Core/Repository/URLWildcardService.php:
```diff
@@ -125,19 +118,31 @@ public function create($sourceUrl, $destinationUrl, $forward = false)
      * @return string
      */
-    protected function cleanUrl($url)
+    protected function cleanUrl($url): string
```
```suggestion
    private function cleanUrl($url): string
```

andrerom replied:
And maybe also :)

```suggestion
    private function cleanUrl(string $url): string
```

alongosz commented on eZ/Bundle/EzPublishCoreBundle/DependencyInjection/Compiler/UrlWildcardRouterPass.php:
```diff
@@ -0,0 +1,48 @@
+            return;
+        }
+
+        if ($container->getParameter('ezpublish.url_wildcards.enabled') === true) {
```
TBH I have strong reservation about this entire compiler pass. I don't really see a point in parametrizing container building based on some settings. It's difficult to maintain as our service definition exist in Yaml files.
Why do we even need to make it optional?

andrerom replied:
Performance overhead  for those not using it (extra lookup for every request) and there is no UI to deal with this in 2.5 _(yet)_ so it's more for migration feature parity for now.

kmadejski replied:
@alongosz, as @andrerom explained, it's meant to be an _optional_ feature just for the customers who migrated from eZ Publish and still have to use UrlWildcards defined in the past (for SEO / migration reasons for instance). It was discussed with PMs and @lserwatka.

alongosz commented on eZ/Bundle/EzPublishCoreBundle/DependencyInjection/EzPublishCoreExtension.php:
```diff
@@ -661,4 +662,13 @@ private function prependTranslatorConfiguration(ContainerBuilder $container)
+            $config['url_wildcards']['enabled'] = false;
+        }
+
+        $container->setParameter('ezpublish.url_wildcards.enabled', $config['url_wildcards']['enabled']);
```
```suggestion
        $container->setParameter('ezpublish.url_wildcards.enabled', $config['url_wildcards']['enabled'] ?? false);
```
and you can drop lines 668-671

alongosz commented on eZ/Publish/Core/MVC/Symfony/Routing/UrlWildcardRouter.php:
```diff
@@ -0,0 +1,169 @@
+     */
+    public function supports($name): bool
+    {
+        return $name === self::URL_ALIAS_ROUTE_NAME;
```
If you want this entire feature to be enabled/disabled, then at the first glance this seems the right place to do it, instead of some container hacking via compiler passes :)
Still don't see a point though.

kmadejski replied:
Not really. This way, I would have to do an additional `if` which leads to (tiny, but still, please keep in mind that this piece of code will be executed on **every** request) additional load. 
I think it's better to not define the service at all if the feature it's not enabled.

alongosz replied:
Defining service dynamically leads to a lot of magic difficult to track.
I'm really not following issue with additional overhead here.
Something like:
```php
return $isEnabled && $name === self::URL_ALIAS_ROUTE_NAME;
```
This way lookup done by this router is not executed unless feature is enabled.

kmadejski replied:
I found a better way to do so, done in: https://github.com/ezsystems/ezpublish-kernel/pull/2609/commits/5c923c1750b5c1973b64cacca5540a888045f35b

alongosz commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlWildcard/Gateway/DoctrineDatabase.php:
```diff
@@ -162,4 +162,33 @@ public function loadUrlWildcardsData($offset = 0, $limit = -1)
+     */
+    public function loadUrlWildcardBySourceUrl(string $sourceUrl)
+    {
+        $query = $this->dbHandler->createSelectQuery();
```
Please use `Doctrine\Connection` to build queries, this will be gone in the next major (if time allows :P)

kmadejski replied:
No, it's not possible. I would have to:
a) refactor the whole class to use `Doctrine\Connection`, which is far away from the scope of this PR,
b) have to inject both `\eZ\Publish\Core\Persistence\Database\DatabaseHandler` and `Doctrine\Connection` in the same time. I'm pretty sure that you wouldn't accept this approach afterward 😉

alongosz replied:
> No, it's not possible. I would have to:
> a) refactor the whole class to use `Doctrine\Connection`, which is far away from the scope of this PR,
> b) have to inject both `\eZ\Publish\Core\Persistence\Database\DatabaseHandler` and `Doctrine\Connection` in the same time. I'm pretty sure that you won't accept this approach afterward

@kmadejski DatabaseHandler will be gone in 3.0/8.0, so this will stop working. I'm not accepting technical debt at this point. Please see how you can achieve something in the middle - using Connection, but not injecting it and not refactoring changes out of scope. It is possible.

kmadejski replied:
@alongosz done in https://github.com/ezsystems/ezpublish-kernel/pull/2609/commits/caaa05b32e11188af3ff5e91ce23845b8ad3ccdd.

alongosz commented:
Side:
> **Bug/Improvement** | feature parity

This is not a correct answer to that question, all available options are given in the template :P

andrerom commented:
> This is not a correct answer to that question, all available options are given in the template :P

The world is not binary @alongosz ;)

alongosz commented:
> > This is not a correct answer to that question, all available options are given in the template :P
> 
> The world outside your computer is not binary @alongosz ;)

:scream:

andrerom commented on eZ/Publish/Core/Persistence/Cache/UrlWildcardHandler.php:
```diff
@@ -0,0 +1,146 @@
+    {
+        $this->logger->logCall(__METHOD__, ['id' => $id]);
+
+        $urlWildcard = $this->load($id);
```
We don't normally load in SPI on operations, any particular reason why you do it here? _(code below does not need it afaik)_

andrerom commented on eZ/Publish/Core/Persistence/Cache/UrlWildcardHandler.php:
```diff
@@ -0,0 +1,146 @@
+     */
+    public function load($id)
+    {
+        $this->logger->logCall(__METHOD__, ['id' => $id]);
```
Wrong use here, `logCall ` is for non cached methods.

Should rather be something like:
```php
 if ($cacheItem->isHit()) {
     $this->logger->logCacheHit(['id' => $id]);

    return $cacheItem->get();
}

$this->logger->logCacheMiss(['id' => $id]);
```

andrerom commented on eZ/Publish/Core/Persistence/Cache/UrlWildcardHandler.php:
```diff
@@ -0,0 +1,146 @@
+     */
+    public function translate(string $sourceUrl): UrlWildcard
+    {
+        $this->logger->logCall(__METHOD__, ['url' => $sourceUrl]);
```
same here

alongosz commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlWildcard/Gateway/DoctrineDatabase.php:
```diff
@@ -162,4 +163,36 @@ public function loadUrlWildcardsData($offset = 0, $limit = -1)
+
+        $result = $queryBuilder->execute()->fetch(FetchMode::ASSOCIATIVE);
+
+        return $result ?: [];
```
Can we use here:
```suggestion
        return $result ?? [];
```
?

kmadejski replied:
Unfortunately, not. This method is supposed to return `array` in all cases, but `$queryBuilder->execute()->fetch()` may return `false`, which is consider as correct for *null* coalescing operator. See: https://3v4l.org/nQOaR

alongosz replied:
Ah, that's true. I'm in such cases fan of something more explicit like 
```php
$result !== false ? $result : []
```` 
but what you have here is good anyway :+1:

alongosz commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlWildcard/Handler.php:
```diff
@@ -113,4 +116,107 @@ public function loadAll($offset = 0, $limit = -1)
+     */
+    private function substitute(string $destinationUrl, array $values): string
+    {
+        preg_match_all(self::PLACEHOLDERS_REGEXP, $destinationUrl, $matches);
```
if `preg_match_all` doesn't match anything it will return `FALSE` and the next statement will trigger `E_NOTICE` warning.
I'm not sure however what is the correct way to handle it. Return `$destinationUrl` anyway or throw RuntimeException. Given that you don't seem to expect that due to current usage, I'd say it should throw.

kmadejski replied:
It's not correct. `preg_match_all` for this particular regexp will **always** return a two-dimensional array (ref: https://3v4l.org/jHXIh), even though it is empty. It might be empty if your destination URL leads to another domain. Therefore, `E_NOTICE` will never be triggered.

alongosz replied:
> It's not correct. `preg_match_all` for this particular regexp will **always** return a two-dimensional array (ref: https://3v4l.org/jHXIh), even though it is empty. It might be empty if your destination URL leads to another domain. Therefore, `E_NOTICE` will never be triggered.

I know. Right now. This is why I approved the PR instead of requesting changes ;) 

It's rather about the fact that this construct, w/o sanity check is not a good pattern, looking entirely at this single method. What if regexp changes? What if the change will result in `E_NOTICE` only in edge cases? If function documents that it can return false and next lines of code depend on the expected execution of that code, it's always good idea to check that. Even if at the time of writing it, it won't occur.

Given this is not that important right now, you can keep it as-is.

mateuszbieniek commented on eZ/Bundle/EzPublishCoreBundle/Resources/config/url_wildcard.yml:
```diff
@@ -0,0 +1,12 @@
+        calls:
+          - [setLogger, ['@?logger']]
+        tags:
+          - { name: router, priority: 210 }
```
Just a nitpick: no newline at the end of the file.

mateuszbieniek commented on eZ/Publish/Core/MVC/Symfony/Routing/UrlWildcardRouter.php:
```diff
@@ -0,0 +1,191 @@
+     */
+    public function getContext(): RequestContext
+    {
+        $this->requestContext;
```
Missing `return`

kmadejski commented:
I fixed case found by @mnocon during QA phase: missing SA in URI on redirect. To do so, I simplified redirection in https://github.com/ezsystems/ezpublish-kernel/pull/2609/commits/6e0cf977ef06df4da7f4ac74e049c43ac9f8ff54.

kmadejski commented:
@alongosz / @andrerom I think we can finally merge this one 🙂

alongosz commented:
Merged & merged up to master.

Thank you @kmadejski. Great work! :tada:

SylvainGuittard commented:
Amazing! You did it Kamil! 🎉
